pilot/pkg/model/authorization.go

```diff
@@ -65,43 +65,80 @@ type AuthorizationPoliciesResult struct {
 	Audit  []AuthorizationPolicy
 }
 
+type ProxyInfo struct {
+	Namespace       string
+	WorkloadName    string
+	IsWaypointProxy bool
+	Workload        labels.Instance
+}
+
+// namespace string, workloadName string, isWaypointProxy bool, workload labels.Instance
 // ListAuthorizationPolicies returns authorization policies applied to the workload in the given namespace.
-func (policy *AuthorizationPolicies) ListAuthorizationPolicies(namespace string, workload labels.Instance) AuthorizationPoliciesResult {
+func (policy *AuthorizationPolicies) ListAuthorizationPolicies(proxyInfo ProxyInfo) AuthorizationPoliciesResult {
 	ret := AuthorizationPoliciesResult{}
 	if policy == nil {
 		return ret
 	}
 
-	var namespaces []string
-	if policy.RootNamespace != "" {
-		namespaces = append(namespaces, policy.RootNamespace)
-	}
-	// To prevent duplicate policies in case root namespace equals proxy's namespace.
-	if namespace != policy.RootNamespace {
-		namespaces = append(namespaces, namespace)
-	}
-
-	for _, ns := range namespaces {
-		for _, config := range policy.NamespaceToPolicies[ns] {
+	if proxyInfo.IsWaypointProxy {
+		for _, config := range policy.NamespaceToPolicies[proxyInfo.Namespace] {
 			spec := config.Spec
-			selector := labels.Instance(spec.GetSelector().GetMatchLabels())
-			if selector.SubsetOf(workload) {
-				switch config.Spec.GetAction() {
-				case authpb.AuthorizationPolicy_ALLOW:
-					ret.Allow = append(ret.Allow, config)
-				case authpb.AuthorizationPolicy_DENY:
-					ret.Deny = append(ret.Deny, config)
-				case authpb.AuthorizationPolicy_AUDIT:
-					ret.Audit = append(ret.Audit, config)
-				case authpb.AuthorizationPolicy_CUSTOM:
-					ret.Custom = append(ret.Custom, config)
-				default:
-					log.Errorf("ignored authorization policy %s.%s with unsupported action: %s",
-						config.Namespace, config.Name, config.Spec.GetAction())
+			targetRef := spec.GetTargetRef()
+
+			// At this time, policies without a targetRef will default to workload selectors
+			// TODO: remove this logic when workloadselector for waypoints is disallowed
+			if targetRef == nil {
+				selector := labels.Instance(spec.GetSelector().GetMatchLabels())
```

Comment:
We could consider adding a log line here to indicate the target is a waypoint proxy, but does not have a `targetRef`.

Reply:
Good catch! I added the log line.